core-bundle/src/Resources/contao/widgets/SerpPreview.php

```diff
@@ -100,87 +105,89 @@ private function getDescription(Model $model)
 		return $model->{$this->serpPreview['description']};
 	}
 
+	/**
+	 * @todo Use the router to generate the URL in a future version (see #831)
+	 */
 	private function getUrl(Model $model)
 	{
-		if (isset($this->serpPreview['url']))
+		if (!isset($this->serpPreview['url']))
 		{
-			return $this->serpPreview['url'];
+			throw new \Exception('Please provide the SERP widget URL as string or callable');
 		}
 
-		// FIXME: use the router to generate the URL (see #831)
-		switch (true)
+		if (\is_callable($this->serpPreview['url']))
 		{
-			case $model instanceof PageModel:
-				return $model->getAbsoluteUrl();
-
-			case $model instanceof NewsModel:
-				return News::generateNewsUrl($model, false, true);
+			$placeholder = bin2hex(random_bytes(10));
 
-			case $model instanceof CalendarEventsModel:
-				return Events::generateEventUrl($model, true);
+			// Pass a detached clone with the alias set to the placeholder
+			$tempModel = clone $model;
+			$tempModel->origAlias = $tempModel->alias;
+			$tempModel->alias = $placeholder;
+			$tempModel->preventSaving(false);
 
-			default:
-				throw new \RuntimeException(sprintf('Unsupported model class "%s"', \get_class($model)));
+			return str_replace($placeholder, '%s', $this->serpPreview['url']($tempModel));
```

Comment:
This implementation means I can‘t use services and DI for my callback. Should‘t we apply the usual `System::importStatic()` if the callable is an array?

Reply:
You are right. Fixed in d31c5ee0fe71cb6654b8c29b6bcbbad5dbe85214.